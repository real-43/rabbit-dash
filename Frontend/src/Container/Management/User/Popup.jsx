import { useState } from "react"

function popup() {

    const [newName, setNewName] = useState(props.currentName)
    const [newPassword, setNewPassword] = useState(props.currentPassword)
    const [role, setRole] = useState(props.currentRole)
    const [isActive, setActive] = useState(false)
    const [isOpen, setIsOpen] = useState(props.isOpen)

    // To show/unshow pass in edit popup
    function displayOption() {
        setActive(!isActive);
        var x = document.getElementById("myInput");

        if (x.type === "password") {
            x.type = "text";
        } else {
            x.type = "password";
        }
    }
    
    return (props.isOpen) ? (
      <div>
        <Modal show={true} onHide={(e)=>{props.onClose}}>
          <Modal.Header>
            <Modal.Title>Edit User<i onClick={(e) => {props.onClose}} style={{cursor:"pointer", marginLeft:"330px"}} className='fa fa-times'/></Modal.Title>
          </Modal.Header>
          <Modal.Body>
            <Form>
              <Form.Group className="mb-3" controlId="exampleForm.ControlInput1">
                <Form.Label>Name</Form.Label>
                <Form.Control
                  type="text"
                  placeholder={props.changeUserName}
                  value={newName}
                  onChange={(e) => {props.nameChange}}
                  autoFocus
                />
              </Form.Group>
              <Form.Group
                className="mb-3"
                controlId="exampleForm.ControlTextarea1"
              >
                <Form.Label>Password</Form.Label>
                <InputGroup className="mb-3">
                  
                  <FormControl
                    aria-label="Example text with button addon"
                    aria-describedby="basic-addon1"
                    id='myInput'
                    type="password"
                    placeholder="password"
                    value={newPassword}
                    onChange={(e) => props.passChange}
                  />
                  <Button variant="outline-secondary" id="button-addon1" onClick={(e) => displayOption(this)}>
                    <i class={isActive ? "fa fa-eye" : "fa fa-eye-slash"} id="togglePassword"/>
                  </Button>
                </InputGroup>
              </Form.Group>
              <Form.Label>Role</Form.Label>
              <Form.Select aria-label={role} defaultValue={role} onChange={(e) => setRole(e.target.value)}>
                <option className="d-none" value="">{role}</option>
                <option value="">Set No role</option>
                {roles.filter(obj => checkDomainRole(obj)).map((role,i) => {return (
                  <option value={role.name} key={i}>{role.name}</option>
                )})}
              </Form.Select>
            </Form>
          </Modal.Body>
          <Modal.Footer>
            <Button variant="secondary" onClick={(e) => {props.onClose}}>
              Close
            </Button>
            <Button variant="primary" onClick={(e) =>{handleChange(changeUser)}}>
              Save Changes
            </Button>
          </Modal.Footer>
        </Modal>
      </div>
    ) : ""
  }