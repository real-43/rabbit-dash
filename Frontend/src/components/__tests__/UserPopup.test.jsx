import React  from "react";
import { Provider } from 'react-redux';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import Popup from "../../Container/Management/User/Components/EditPopup"
import Adapter from '@wojtekmaj/enzyme-adapter-react-17'
import { mount, shallow, configure } from 'enzyme';
import { render, screen } from '@testing-library/react';
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk';
import '@testing-library/jest-dom'

configure({adapter: new Adapter()});

const mockStore = configureMockStore([thunk]);

const store = mockStore({
    Management: {
      Permission: ["Datafile Monitoring","BTS QR"],
      Project: ["Datafile Monitoring","BTS QR"], 
      Services: ["Datafile Monitoring","BTS QR"]
    },
    name:"Admin",
    project: [
        {name:"Datafile Monitoring",subMenu:["1","2"]},
        {name:"BTS QR",subMenu:["1","2"]}
    ]
  },
  {
    Management: {
      Permission: ["Datafile Monitoring"],
      Project: ["Datafile Monitoring"], 
      Services: ["Datafile Monitoring"]
    },
    name:"DatafileMonitoringAdmin",
    project: [{name:"Datafile Monitoring",subMenu:["1","2"]}]
  }
);

const currentRole = {
    Management: {
        Permission: ["Datafile Monitoring"],
        Project: ["Datafile Monitoring"], 
        Services: ["Datafile Monitoring"]
    },
    name:"DatafileMonitoringAdmin",
    project: [{name:"Datafile Monitoring",subMenu:["1","2"]}]    
}

const changeUser = {
    id: "Lw8rguKkI9WorgC8M7m9f5TuObg2", 
    email:"test@rabbit.co.th",
    isBlocked: false,
    name: "test",
    password: "123123",
    role: "DatafileMonitoringStaff"
}

const currentUser = {
    id: "Lw8rguKkI9WorgC8M7m9f5TuObg2", 
    email:"test@rabbit.co.th",
    isBlocked: false,
    name: "test",
    password: "123123",
    role: "DatafileMonitoringStaff"
}

describe('renders content when modal is open', () => {
    // mount renders to the dom (real or mocked)
    
    test("render Save button", () => {
        const {asFragment, getByText} = render(
                <Provider store={store} >
                <BrowserRouter styles={{pointerEvents: 'none',cursor: 'none'}}>
                    <Routes>
                        <Route path='/' exact element={<Popup 
                            roles={Object.values(store)}
                            changeUser={changeUser} 
                            newName={'newName'}
                            newPassword={'newPassword'}
                            currentUser={currentUser}
                            currentRole={currentRole}
                            role={"DatafileMonitoringStaff"}
                            onClose={() => {setIsOpen(false)}}
                        />} />
                    </Routes>
                </BrowserRouter>
                </Provider>
        );
        const primaryButton = screen.getByRole('button', { name: /Save Change/i })
        expect(primaryButton).toHaveClass('btn-primary')
    })
    test("render Close  button", () => {
        const {asFragment, getByText} = render(
                <Provider store={store} >
                <BrowserRouter styles={{pointerEvents: 'none',cursor: 'none'}}>
                    <Routes>
                        <Route path='/' exact element={<Popup 
                            roles={Object.values(store)}
                            changeUser={changeUser} 
                            newName={'newName'}
                            newPassword={'newPassword'}
                            currentUser={currentUser}
                            currentRole={currentRole}
                            role={"DatafileMonitoringStaff"}
                            onClose={() => {setIsOpen(false)}}
                        />} />
                    </Routes>
                </BrowserRouter>
                </Provider>
        );
        const secondaryButton = screen.getByRole('button', { name: /Close/i })
        expect(secondaryButton).toHaveClass('btn-secondary')
    })
    test("render name input", () => {
        const {asFragment, getByText} = render(
                <Provider store={store} >
                <BrowserRouter styles={{pointerEvents: 'none',cursor: 'none'}}>
                    <Routes>
                        <Route path='/' exact element={<Popup 
                            roles={Object.values(store)}
                            changeUser={changeUser} 
                            newName={'newName'}
                            newPassword={'newPassword'}
                            currentUser={currentUser}
                            currentRole={currentRole}
                            role={"DatafileMonitoringStaff"}
                            onClose={() => {setIsOpen(false)}}
                        />} />
                    </Routes>
                </BrowserRouter>
                </Provider>
        );
        const passInput = screen.getByRole('textbox', { value: /Name/i })
        expect(passInput).toHaveClass('form-control')
    })   
});
describe('render Popup component correctly', () => {  
        const Popup = shallow(
        <Provider store={store}>
        <BrowserRouter styles={{pointerEvents: 'none',cursor: 'none'}}>
            <Routes>
                <Route path='/' exact element={
                    <Popup/>
                } />
            </Routes>
        </BrowserRouter>
        </Provider>
    ).dive();
    expect(Popup).toMatchSnapshot();
});
