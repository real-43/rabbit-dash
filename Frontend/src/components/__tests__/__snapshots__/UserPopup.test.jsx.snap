// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `ShallowWrapper {}`;

exports[`render Popup component correctly 1`] = `ShallowWrapper {}`;
